### PR TITLE
First commit

### PR DESCRIPTION
**Pull request by basiliscos**

### CODE REVIEW

File: lib/Mojolicious/Plugin/ClientIP/Pluggable.pm [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+sub _check_ipv4 {
+    my ($ip) = @_;
+    return !Data::Validate::IP::is_unroutable_ipv4($ip)
+        && !Data::Validate::IP::is_private_ipv4($ip)
+        && (SKIP_LOOPBACK || !Data::Validate::IP::is_loopback_ipv4($ip));
+}
+
+sub _check_ipv6 {
+    my ($ip) = @_;
+    return !Data::Validate::IP::is_public_ipv4($ip);
```
Comment (aminmarashi) [inline lib/Mojolicious/Plugin/ClientIP/Pluggable.pm]: Just curious, why didn't use `is_public_ipv6 `?

File: lib/Mojolicious/Plugin/ClientIP/Pluggable.pm [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+use Mojo::Base 'Mojolicious::Plugin';
+
+our $VERSION = '0.01';
+
+# for tests only
+use constant SKIP_LOOPBACK => $ENV{CLIENTIP_PLUGGABLE_SKIP_LOOPBACK} || 1;
+
+sub _check_ipv4 {
+    my ($ip) = @_;
+    return !Data::Validate::IP::is_unroutable_ipv4($ip)
```
Comment (aminmarashi) [inline lib/Mojolicious/Plugin/ClientIP/Pluggable.pm]: Why not `is_public_ipv4`?

File: lib/Mojolicious/Plugin/ClientIP/Pluggable.pm (hunk 1 above)
Comment (basiliscos) [reply lib/Mojolicious/Plugin/ClientIP/Pluggable.pm]: because it is bug :)

File: lib/Mojolicious/Plugin/ClientIP/Pluggable.pm (hunk 2 above)
Comment (basiliscos) [reply lib/Mojolicious/Plugin/ClientIP/Pluggable.pm]: Actually I'm not sure now, that `is_public_*` is correct methods, as it fails to validate `203.0.113.43`.

Comment (aminmarashi) [reply lib/Mojolicious/Plugin/ClientIP/Pluggable.pm]: @basiliscos I think because of https://tools.ietf.org/html/rfc5735

Comment (basiliscos) [reply lib/Mojolicious/Plugin/ClientIP/Pluggable.pm]: Indeed! Thanks! I fixed the code & tests.

File: lib/Mojolicious/Plugin/ClientIP/Pluggable.pm [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+sub _classify_ip {
+    my ($ip) = @_;
+    return Data::Validate::IP::is_ipv4($ip)
+        ? 'ipv4'
+        : Data::Validate::IP::is_ipv6($ip)
+        ? 'ipv6'
+        : undef;
+}
+
+sub _candatetes_iterator {
```
Comment (raunakkathuria) [inline lib/Mojolicious/Plugin/ClientIP/Pluggable.pm]: may be rename it to better `_candidates_iterator`, i don't know what `candatetes` means and i don't think there is work like this

File: lib/Mojolicious/Plugin/ClientIP/Pluggable.pm [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+Mojolicious::Plugin::ClientIP::Pluggable is a Mojolicious plugin to get an IP address, which
+allows to specify different HTTP-headers (and their priporities) for client IP address
+extraction. This is needed as different cloud providers set different headers to disclose
+real IP address.
+
+If the address cannot be extracted from headers different fallback options are available:
+detect IP address from C<X-Forwarded-For> header, detect IP address from C<forwarded> header
+(rfc-7239), or use C<remote_address> environment.
+
+The pluging is inspired by L<Mojolicious::Plugin::ClientIP>.
```
Comment (tom-binary) [inline lib/Mojolicious/Plugin/ClientIP/Pluggable.pm]: `plugin`

File: lib/Mojolicious/Plugin/ClientIP/Pluggable.pm [hunk 5]
```diff
@@ -0,0 +1,215 @@
+package Mojolicious::Plugin::ClientIP::Pluggable;
+
+=head1 NAME
+
+Mojolicious::Plugin::ClientIP::Pluggable - Customizable way for client IP detection plugin
```
Comment (tom-binary) [inline lib/Mojolicious/Plugin/ClientIP/Pluggable.pm]: `Customizable client IP detection plugin for Mojolicious`

File: lib/Mojolicious/Plugin/ClientIP/Pluggable.pm [hunk 6]
```diff
@@ -0,0 +1,215 @@
+package Mojolicious::Plugin::ClientIP::Pluggable;
+
```
Comment (tom-binary) [inline lib/Mojolicious/Plugin/ClientIP/Pluggable.pm:2]: needs an `# ABSTRACT: ...` line as well, I'd suggest something like:

```
# ABSTRACT: Client IP header handling for Mojolicious requests
```

File: lib/Mojolicious/Plugin/ClientIP/Pluggable.pm [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+        my $c = shift;
+        $c->render(text => $c->client_ip);
+    };
+
+    app->start;
+
+=head1 DESCRIPTION
+
+Mojolicious::Plugin::ClientIP::Pluggable is a Mojolicious plugin to get an IP address, which
+allows to specify different HTTP-headers (and their priporities) for client IP address
```
Comment (tom-binary) [inline lib/Mojolicious/Plugin/ClientIP/Pluggable.pm]: `priorities`

File: lib/Mojolicious/Plugin/ClientIP/Pluggable.pm [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+            next unless $address_family;
+
+            # possibly limit to acceptable address family
+            next if $restrict_family && $restrict_family ne $address_family;
+
+            # validate by family
+            my $validator = $validator_for{$address_family};
+            next unless $validator->($ip);
+
+            # address seems valid, return it's textual representation
```
Comment (tom-binary) [inline lib/Mojolicious/Plugin/ClientIP/Pluggable.pm]: `its`

**Pull request merged**